compiler/luci/pass/src/FoldShapePass.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldShapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+template <loco::DataType OutType> luci::CircleConst *shape_of(luci::CircleNode *input_tensor)
+{
+  auto name = input_tensor->name();
+  auto shape_status = input_tensor->shape_status();
+  auto rank = input_tensor->rank();
+  assert(name.length() > 0);
+  assert(shape_status == luci::ShapeStatus::VALID);
+  assert(rank > 0);
+
+  auto shape = input_tensor->graph()->nodes()->create<luci::CircleConst>();
+  shape->name(name + "_ConstShape");
+  shape->dtype(OutType);
+  shape->rank(1);
+  shape->dim(0).set(rank);
+
+  shape->size<OutType>(rank);
+  for (uint32_t i = 0; i < rank; i++)
+    shape->at<OutType>(i) = input_tensor->dim(i).value();
+
+  return shape;
+}
+
+/**
+ * Fold Shape to const if the input shape is fixed
+ **/
+template <loco::DataType OutType> bool fold_shape(luci::CircleShape *shape)
+{
+  auto input_tensor = dynamic_cast<luci::CircleNode *>(shape->input());
```

Comment:
1. You can use `luci::must_cast` function here as the input must be dynamically casted.
2. I added a const specifier.
3. The name might as well include "node" because it is a circle node. This change will affect following variables as well.
```suggestion
  auto input_node = loco::must_cast<const luci::CircleNode *>(shape->input());
```

Reply:
1. I refered other folding passes (e.g. `FoldGatherPass` / `FoldCastPass` / etc... ) and they used `dynamic_cast`
May I ask why you recommend to use `luci::must_cast`? 😃

2. If I add `const` specifier before `luci::CircleNode`, it conflicts at line 33
```cpp
auto shape = input_tensor->graph()->nodes()->create<luci::CircleConst>();
```

Reply:
> May I ask why you recommend to use luci::must_cast? 😃

We casts dynamically the node if the node could be some type. And, we `must_cast` the node if it must be some type.

For example, let's say you're trying to dynamically cast `CircleNode` into `CircleRelu` because the pass only takes `CircleRelu` as an input. You will traverse all the nodes and the result of `dynamic_cast` could be `nullptr` because it is not `CircleRelu`.

But, above case it about casting `shape->input()`. It must be `CircleNode`. There is no need to check if it is `nullptr`.

> it conflicts at line 33

Ah, okay.